klient: rework logging

rjeczalik opened the pull request:
~~deps: https://github.com/koding/koding/pull/8904~~

This PR reworks logging in klient, making it possible to upload arbitrary log files from remote machines using keygen and logrotate packages. Please take a look.

This PR also adds "log.upload" method to klient kite, which allows for adding e.g. `kd upload` command if needed. /cc @leeola

rjeczalik committed:
gatewaytest: refactor out a common testing package

rjeczalik committed:
logrotate: add package for streaming & rotating log files

rjeczalik committed:
logrotate: do not hide NopError in Upload

rjeczalik committed:
kites/keygen: rename gateway -> keygen

rjeczalik committed:
logrotate: fix counting part sizes

rjeczalik committed:
s3logrotate: remove package, superseded by koding/logrotate

rjeczalik committed:
logrotate: rename Logger -> Uploader

rjeczalik committed:
klient: integrate logrotate.Uploader and keygen.UserBucket

rjeczalik committed:
kloud/provider: make user-data script user-accessible

rjeczalik committed:
koding/config: add bucket and kloud endpoint

rjeczalik committed:
koding: rename gateway env -> keygen

cihangir commented on go/src/koding/kites/keygen/keygentest/gatewaytest.go:
```diff
@@ -1,18 +1,60 @@
-package gateway_test
```
file name still "gateway" intentional?

rjeczalik replied:
done, renamed to keygentest.go, I missed this any many others 😇

cihangir commented on go/src/koding/kites/keygen/keygentest/gatewaytest.go:
```diff
@@ -114,6 +156,29 @@ func (d *Driver) keyPair() *kitetest.KeyPair {
+		Underlying() *Flags
+	}
+
+	envPrefix := "gateway"
```
why do you choose gateway for env vars instead of the service/worker name?

rjeczalik replied:
done, renamed to "keygen"

cihangir commented on go/src/koding/klient/uploader/uploader.go:
```diff
@@ -0,0 +1,275 @@
+	Log       kite.Logger // optional; defaultLog if nil
+}
+
+// Logger is a wrapper for kite.Logger that implements logging.Logger.
```
i didnt understand the requirement here, could you please explain? AFAIK kite.logger is already implements koding.logger as we can see here https://github.com/koding/kite/blob/02ecbfc2262fb6dd7be61a0ff23b7a79431a02fe/logger.go#L80

rjeczalik replied:
The code you linked creates kite.Logger from logging.Logger. Here's the opposite - creating logging.Logger from kite.Logger.

rjeczalik replied:
Now I got it on a second read - the kite.Logger interface can be upgraded to logging.Logger interface, since the underlying value implements logging.Logger when created by kite.

Done, I've removed the wrapper in https://github.com/koding/koding/pull/8947/commits/71eef548cb053017890af6c96fce5f2520b6d159.

rjeczalik committed:
koding: rename gateway -> keygen everywhere

rjeczalik commented on config/generateKonfig.coffee:
```diff
@@ -138,6 +138,10 @@ module.exports = (options, credentials) ->
     userPublicKey: credentials.kloud.userPublicKey
     userPrivateKey: credentials.kloud.userPrivateKey
 
+    keygenAccessKey: credentials.kloud.gatewayAccessKey
```
done

cihangir commented on go/src/koding/klient/uploader/uploader.go:
```diff
@@ -0,0 +1,275 @@
+		var req *UploadRequest
+
+		switch n, v, _ := reflect.Select(cases); n {
+		case 0:
```
why we need to use reflect here?

rjeczalik replied:
Do be able to select on dynamic number of channels.

Alternative would be something like:

```go
tickers := []*time.Ticker

loop:
    for {
        for _, i := range rand.Perm(len(tickers)) {
            switch {
            case <-tickers[i].C:
                // ...
                // handling here
                // ...
                continue loop
            default:
                continue
            }
        }
    }
```

I find the reflect more readable & simpler.

cihangir replied:
i think another way to do that could be

```
	for {
		select {
		case <-up.close:
			return
		case req := <-up.req:
			go func() {
				t := time.NewTicker(req.Interval)
				for {
					select {
					case <-up.close:
						t.Stop()
						return
					case <-t.C:
						// do your magic here
					}
				}
			}()
			// do your magic here, for the first time
		}
	}
```

cihangir replied:
Also please check if the files are already listened/watched by another ticker, multiple requests to kite.upload logs could result in multiple tickers

rjeczalik replied:
Also would work, but I don't like starting additional goroutines for something, that could be done in one. In my opinion it's easier to reason about concurrency and do not worry about leaking goroutines.

Your call, if you find it easier to read I can rework reflect.Select to multiple goroutines. :-)

rjeczalik replied:
> Also please check if the files are already listened/watched by another ticker

done in https://github.com/koding/koding/pull/8947/commits/ee069993f0a3e7b6e11b92b87ff62e35502f9e6c

cihangir replied:
I think we have different takes on readability :) 

Since you said  "your call" lets change it to non-reflected version not to fiddle with the runtime memory 🙋

rjeczalik replied:
done in https://github.com/koding/koding/pull/8947/commits/d59fe509b2e3437bb1d7be208ac35878799f9c56

rjeczalik committed:
klient/uploader: remove Logger wrapper, upgrade to logging.Logger interface

rjeczalik committed:
klient/uploader: do not watch the same file multiple times

rjeczalik committed:
klient/uploader: rework up.process to not use reflect

rjeczalik committed:
config: fix bucket config for keygen

rjeczalik committed:
logrotate: gzip content by default

rjeczalik committed:
klient/uploader: return actual url of the uploaded part

Since the final key is generated by logrotate depending
on the part sequence number, we need to pass the effective
url down from logrotate.

This PR reworks whole call chain to get the url in uploader
package.

leeola commented:
LGTM

rjeczalik commented:
@leeola @cihangir Added one more - https://github.com/koding/koding/commit/d37869569c89391602334e48fda86c38b551f9bc#diff-18cc3d0abd4e3e71be2ae47707313dfeR42. Please take a look.

rjeczalik committed:
klient: fix vagrant template

rjeczalik committed:
logrotate: increase DefaultChecksumSize to 256

In debug mode the are a lot of lines like:

  2016-09-10T07:39:07.65Z  DEBUG    [tunnelclient ][PID:69307][logging/context.go:53] [transport] Opening a new stream from server session

Which are >128 characters. This makes logrotate errornously
no rotating log files. Increasing the checksum size 256
fixes the problem.

leeola commented:
:+1:, LGTM
